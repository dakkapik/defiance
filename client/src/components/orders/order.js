const Orderdata = [
  {
    _id: {
      $oid: "5fe36229a00b6286d0a20673",
    },
    status: "unassigned",
    orderNumber: 270,
    date: "10/07",
    time: "05:41pm",
    firstName: "KATHY",
    lastName: "na",
    phoneNumber: "(954) 599-4250",
    address: "6708 NW 22ND",
    city: "MARGATE",
    zip: "33073",
    __v: 0,
  },
  {
    _id: {
      $oid: "5fe3623fa00b6286d0a20674",
    },
    status: "unassigned",
    orderNumber: 272,
    date: "10/07",
    time: "05:43pm",
    firstName: "na",
    lastName: "KARA",
    phoneNumber: "(954) 422-7494",
    address: "6153 NW 41ST DRIVE",
    city: "Coral Springs",
    zip: "33067",
    __v: 0,
  },
  {
    _id: {
      $oid: "5fe36251a00b6286d0a20675",
    },
    status: "unassigned",
    orderNumber: 273,
    date: "10/07",
    time: "05:44pm",
    firstName: "LUIS",
    lastName: "LUIS",
    phoneNumber: "(240) 813-5918",
    address: "4691 NW 66TH DRIVE",
    city: "CORAL SPRINGS",
    zip: "33073",
    __v: 0,
  },
];

let ParseOrders = {
  Orders: {},
  columns: {
    "column-1": {
      id: "column-1",
      title: "Orders",
      OrderIds: [],
    },
  },
  columnOrder: [],
};

console.log(ParseOrders);

// Orderdata.map((e, i) => console.log(e));

export default Orderdata;
