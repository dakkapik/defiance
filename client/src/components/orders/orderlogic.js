const presistTaskFirstColumn = (
  startTaskId,
  source,
  destination,
  draggableId,
  start
) => {
  const newTaskIds = Array.from(startTaskId);

  newTaskIds.splice(source.index, 1);
  newTaskIds.splice(destination.index, 0, draggableId);
  const newColumn = {
    ...start,
    orderIds: newTaskIds,
  };
  const newState = {
    ...dragdropdata,
    columns: {
      ...dragdropdata.columns,
      [newColumn.id]: newColumn,
    },
  };
  setDragDropData(newState);
  return;
};
const presistTaskAllColumns = (
  startTaskId,
  finishTaskId,
  source,
  destination,
  draggableId,
  start,
  finish
) => {
  //When a task move to a different columns
  const startTaskIds = Array.from(startTaskId);
  startTaskIds.splice(source.index, 1);
  const newStart = {
    ...start,
    orderIds: startTaskIds,
  };
  const finishTaskIds = Array.from(finishTaskId);
  finishTaskIds.splice(destination.index, 0, draggableId);
  const newFinish = {
    ...finish,
    orderIds: finishTaskIds,
  };
  const newState = {
    ...dragdropdata,
    columns: {
      ...dragdropdata.columns,
      [newStart.id]: newStart,
      [newFinish.id]: newFinish,
    },
  };
  setDragDropData(newState);
};
