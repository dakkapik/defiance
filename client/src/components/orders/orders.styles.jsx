import styled, { keyframes } from "styled-components";

export const animationName = keyframes`
  0% {
    opacity: 0;
  }
  100% {
    opacity: 1;
  }
}
`;

export const Container = styled.div`
  width: 70%;
  margin: auto;
  justify-content: center;
  display: flex;
  animation: ${animationName} ease 3s;
`;

export const ContainerOrder = styled.div`
  height: 100%;
  width: 50%;
  margin: 1vh;
`;

export const ContainerDriver = styled.div`
  display: flex;
  flex-direction: column;
  width: 50%;
  margin: 1vh;
`;
