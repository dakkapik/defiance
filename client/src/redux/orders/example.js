const Data = {
  tasks: {
    "task-1": { id: "task-1", content: "i am task-1" },
    "task-2": { id: "task-2", content: "i am task-2" },
    "task-3": { id: "task-3", content: "i am task-3" },
    "task-4": { id: "task-4", content: "i am task-4" },
  },

  columns: {
    "column-1": {
      id: "column-1",
      title: "Orders",
      taskIds: ["task-1", "task-2", "task-3", "task-4"],
    },
    8954: {
      id: "8954",
      title: "Driver 8954",
      taskIds: [],
    },
    2342: {
      id: "2342",
      title: "Driver 2342",
      taskIds: [],
    },
  },

  columnOrder: ["column-1", "8954", "2342"],
};
export default Data;
