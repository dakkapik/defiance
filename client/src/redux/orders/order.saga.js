const getSTATE_FROM_DRIVER_REDUCER = (state) => state.drivers.justadded;
/*
 let ParseOrders = {
      orders: {},
      columns: {
        "column-1": {
          id: "column-1",
          title: "Orders",
          orderIds: [],
        },
      },
      columnOrder: ["column-1"],
    };
*/

export function* RETREIVE_DATA_FROM_LISTENER() {
  const items = yield select(getSTATE_FROM_DRIVER_REDUCER);

  console.log("ADD ACTIVE DRIVER IN STORE SAGA", items);
}

export function* OnActiveDriverStart() {
  yield takeLatest(DriversActionTypes.ADD_ACTIVE_DRIVER, tester);
}

export function* driversSagas() {
  yield all([call(onDriverSocketStart)]);
}
