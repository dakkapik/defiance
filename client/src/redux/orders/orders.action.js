import OrdersActionTypes from "./orders.types";

//  when api order succeeds add it to currentdragdrop reducer
export const addApiOrderSuccessDragDrop = (OrdersAndStore) => ({
  type: OrdersActionTypes.ADD_APIORDER_SUCCESS_DRAG_DROP_TO_COLLECTION,
  payload: OrdersAndStore,
});
//when socket gives an order update
export const socketOrderDisplayUpdate = (driver_completed_order) => ({
  type: OrdersActionTypes.ORDER_DISPLAY_SOCKET_UPDATE,
  payload: driver_completed_order,
});

// when api order fails add custom failure to currentdragdrop reducer
export const addApiOrderFailureDragDrop = (OrdersAndStore) => ({
  type: OrdersActionTypes.ADD_APIORDER_FAILURE_DRAG_DROP_TO_COLLECTION,
  payload: OrdersAndStore,
});
// When manager clicks the save button
export const saveOrders = () => ({
  type: OrdersActionTypes.SAVE_ORDER,
});
// When the manager drags an order within the order column
export const persistOrderColumn = (onDragEndProperties) => ({
  type: OrdersActionTypes.PERSIST_ORDER_COLUMN,
  payload: onDragEndProperties,
});
// When the manager drags an order within other columns like drivers
export const persistAllColumn = (onDragEndProperties) => ({
  type: OrdersActionTypes.PERSIST_ALL_COLUMN,
  payload: onDragEndProperties,
});
// check the difference of drivers within state
export const deltaDriverDragAndDrop = (driversAndstore) => ({
  type: OrdersActionTypes.DELTA_DRIVER_FOR_DRAG_AND_DROP,
  payload: driversAndstore,
});
// when a driver leaves within the socket for more detail it's in drivers.utlis in drivers folder
export const removeDriverDragDrop = (driversAndremoveDriver) => ({
  type: OrdersActionTypes.REMOVE_DRIVER_FROM_DRAG_AND_DROP,
  payload: driversAndremoveDriver,
});
//when manager presses x button to an order
export const removeOrderFromDriver = (order) => ({
  type: OrdersActionTypes.REMOVE_ORDER_FROM_DRIVER,
  payload: order,
});

export const discardOrderChanges = () => ({
  type: OrdersActionTypes.DISCARD_ORDER_CHANGES,
});

// UI CASES for expanded the mapsidebar-component and compressing
export const ordersSocketOn = () => ({
  type: OrdersActionTypes.ORDERS_SOCKET_ON,
});
export const ordersSocketOff = (arrow_modal_bool) => ({
  type: OrdersActionTypes.ORDERS_SOCKET_OFF,
  payload: arrow_modal_bool,
});
