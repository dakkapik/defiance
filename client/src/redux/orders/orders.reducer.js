import OrdersActionTypes from "./orders.types";
import {
  addDragDropToCollection,
  getCurrentDragandDrop,
  PresistOrderColumn,
  SaveDragDropCollection,
  initalizeDriverDragDrop,
  PresistAllColumns,
  RemoveDriverFromDragAndDrop,
} from "./orders.utils";
import { differenceBy } from "lodash";
const INITIAL_STATE = {
  showorders: false,
  apiorders: {},
  dragdropcollection: [],
  currentdragdrop: {},
};

const ordersReducer = (state = INITIAL_STATE, action) => {
  switch (action.type) {
    case OrdersActionTypes.ADD_DRAG_DROP_TO_COLLECTION:
      return {
        ...state,
        dragdropcollection: addDragDropToCollection(
          state.dragdropcollection,
          action.payload
        ),
        currentdragdrop: getCurrentDragandDrop(
          state.dragdropcollection,
          action.payload
        ),
      };

    case OrdersActionTypes.PERSIST_ORDER_COLUMN:
      const NewDragDropData = PresistOrderColumn(
        state.currentdragdrop,
        action.payload
      );

      return {
        ...state,
        currentdragdrop: { ...NewDragDropData },
        dragdropcollection: SaveDragDropCollection(
          state.dragdropcollection,
          NewDragDropData
        ),
      };
    case OrdersActionTypes.PERSIST_ALL_COLUMN:
      const NewDragAllDropData = PresistAllColumns(
        state.currentdragdrop,
        action.payload
      );

      return {
        ...state,
        currentdragdrop: { ...NewDragAllDropData },
        dragdropcollection: SaveDragDropCollection(
          state.dragdropcollection,
          NewDragAllDropData
        ),
      };
    /*

    When a manager connects to a store and disconnects and the
    drivers are incoming and leaving  in the background
    Initalize driver will update the drag and drop
    
    */
    case OrdersActionTypes.INITALIZE_DRIVER_FOR_DRAG_AND_DROP:
      //first find drag drop in collection
      const currentDragDrop = state.dragdropcollection.find(
        (collection) => collection.storename === action.payload.storename
      );

      if (currentDragDrop === undefined) return { ...state };

      const newDriver = differenceBy(
        action.payload.driver,
        currentDragDrop.currentdriver,
        "employeeId"
      );

      const RemoveDriver = differenceBy(
        currentDragDrop.currentdriver,
        action.payload.driver,
        "employeeId"
      );

      const NewDragDrop = initalizeDriverDragDrop(
        currentDragDrop,
        newDriver,
        RemoveDriver
      );

      return {
        ...state,

        // if you don't specify a new Object then react components will not update
        currentdragdrop: {
          ...NewDragDrop,
          currentdriver: [...action.payload.driver],
        },
        dragdropcollection: SaveDragDropCollection(state.dragdropcollection, {
          ...NewDragDrop,
          currentdriver: [...action.payload.driver],
        }),
      };
    case OrdersActionTypes.REMOVE_DRIVER_FOR_DRAG_AND_DROP:
      const NewDriver = RemoveDriverFromDragAndDrop(
        state.currentdragdrop,
        action.payload
      );

      return {
        ...state,
        currentdragdrop: { ...NewDriver },
        dragdropcollection: SaveDragDropCollection(state.dragdropcollection, {
          ...NewDriver,
        }),
      };

    //API UPDATES
    case OrdersActionTypes.ORDER_API_SUCCESS:
      return {
        ...state,
        apiorders: action.payload,
      };
    //UI UPDATES
    case OrdersActionTypes.ORDERS_SOCKET_ON:
      return {
        ...state,
        showorders: true,
      };
    case OrdersActionTypes.ORDERS_SOCKET_OFF:
      return {
        ...state,
        showorders: false,
      };

    default:
      return state;
  }
};

export default ordersReducer;
