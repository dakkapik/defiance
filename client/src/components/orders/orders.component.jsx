//external libraries
import React from "react";
import { connect } from "react-redux";
import { DragDropContext } from "react-beautiful-dnd";
//styling
import { Container, ContainerDriver, ContainerOrder } from "./orders.styles";
//actions
import {
  PresistOrderColumn,
  PresistAllColumn,
} from "../../redux/orders/orders.action";

import Column from "../column/column.component";
const Orders = ({ PresistOrderColumn, PresistAllColumn, currentdragdrop }) => {
  const onDragEnd = (result) => {
    const { destination, source, draggableId } = result;
    // if it went out of bounds we don't do anything etheir
    if (!destination) {
      return;
    }
    // if a draggable object stayed at the same position then
    // we don't need to do anything
    if (
      destination.droppableId === source.droppableId &&
      destination.index === source.index
    ) {
      return;
    }

    const start = currentdragdrop.columns[source.droppableId];
    const finish = currentdragdrop.columns[destination.droppableId];

    // If a draggable object remained in the first column
    // we presist the value within the first column
    if (start === finish) {
      PresistOrderColumn({
        startorderids: start.orderIds,
        start: start,
        source: source,
        destination: destination,
        draggableId: draggableId,
      });
      return;
    } else {
      PresistAllColumn({
        startorderids: start.orderIds,
        finishorderids: finish.orderIds,
        source: source,
        destination: destination,
        draggableId: draggableId,
        start: start,
        finish: finish,
      });
    }
  };
  //Array of html elements
  let drivers = [];

  return (
    <div>
      <Container>
        <DragDropContext onDragEnd={onDragEnd}>
          {currentdragdrop.columnOrder.map((columnId, index) => {
            const column = currentdragdrop.columns[columnId];
            // console.log("error", column.orderIds);
            const tasks = column.orderIds.map(
              (orderIds) => currentdragdrop.orders[orderIds]
            );

            if (column.id === "column-1") {
              return (
                <ContainerOrder key={index}>
                  <Column key={column.id} column={column} tasks={tasks} />
                </ContainerOrder>
              );
            } else {
              drivers.push(
                <Column key={column.id} column={column} tasks={tasks} />
              );
            }
            //At the final iteration we want to display the driver in a containerized
            // div
            if (currentdragdrop.columnOrder.length - 1 === index) {
              return (
                <ContainerDriver key={index}>
                  {drivers.map((e, i) => e)}
                </ContainerDriver>
              );
            }
            return null; //placeholder
          })}
        </DragDropContext>
      </Container>
    </div>
  );
};

const mapStateToProps = (state) => ({
  currentdragdrop: state.orders.currentdragdrop,
});

const mapDispatchToProps = (dispatch) => ({
  PresistOrderColumn: (onDragEndProperties) =>
    dispatch(PresistOrderColumn(onDragEndProperties)),
  PresistAllColumn: (onDragEndProperties) =>
    dispatch(PresistAllColumn(onDragEndProperties)),
});

export default connect(mapStateToProps, mapDispatchToProps)(Orders);
