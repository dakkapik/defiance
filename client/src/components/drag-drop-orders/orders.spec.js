import React from "react";
import { Orders } from "./orders.component";
import { shallow } from "enzyme";

describe("Orders component render", () => {
  let wrapper;
  let mockPresistOrderColumn;
  let mockPresistAllColumn;

  beforeEach(() => {
    mockPresistOrderColumn = jest.fn();
    mockPresistAllColumn = jest.fn();
    const mockProps = {
      PresistOrderColumn: mockPresistOrderColumn,
      PresistAllColumn: mockPresistAllColumn,
      currentdragdrop: {
        columnOrder: ["column-1"],
        columns: {
          "column-1": { id: "column-1", orderIds: [], title: "Orders" },
        },
        orders: {},
        storename: "Royal Palm",
      },
    };

    wrapper = shallow(<Orders {...mockProps} />);
  });

  it("should render Orders component", () => {
    expect(wrapper).toMatchSnapshot();
  });
});
