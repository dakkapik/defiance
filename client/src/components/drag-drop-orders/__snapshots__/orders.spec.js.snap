// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Orders component render should render Orders component 1`] = `ShallowWrapper {}`;
